peps/pep-0741.rst

```diff
@@ -1053,6 +1053,89 @@ implemented in Python 3.8.
+-----------------------------------------------
+
+`Eric Snow proposed
+<https://discuss.python.org/t/pep-741-python-configuration-c-api-second-version/45403/27>`_
+to have multiple "phases" for the initalization, not only the
+"preinitialization" and the "initialization". One advantage is to be
+able to configure Python using Python code rather than C code for last
+phases. See :pep:`432` which proposed a similar concept of multiple
+phases, whereas :pep:`587` was implemented only with the
+preinitialization and the initialization.
+
+Conceptually, the runtime can be divided into 5 distinct parts and
+initialization/config may be split along those lines:
+
+* host process resources;
+* global runtime state;
+* interpreter state;
+* thread state;
+* granular components (e.g. multiple interpreters, threading, atexit).
+
+Under the status quo, config/initialization/finalization is currently
+all tangled together. That translates into a consistent maintenance
+cost. The situation is largely a consequence of organic growth over the
+decades (see :pep:`432`), but has gotten better since around 2017.
+
+Eric's ideal equivalent would be something like this:
+
+1. User initializes **host process resources**.
+2. User populates a **configuration**:
+
+   * Set custom values on ``PyConfig``.
+   * ``PyConfig_Update()`` (or something like that).
+     * (some) env vars → config.
+     * host process resources → config.
+   * Set custom values on ``PyInterpreterConfig``.
+
+3. Initialize the runtime
+
+   * ``Py_InitializeFromConfig()``: ``PyConfig`` → **global runtime state**.
+
+4. Create/Initialize the main interpreter
+
+   * ``Py_NewInterpreterFromConfig()`` creates a new interpreter:
+     ``PyInterpreterConfig`` → interpreter state.
+
+5. Set up the main thread:
+
+   * ``PyThreadState_New()``.
+   * ``PyThreadState_Bind()``.
+   * Register host process callbacks (e.g. signal handlers).
+
+6. Enable desired components (e.g. threading).
+7. User starts using the C-API: state → behavior.
+
+``Py_Main()`` would effectively be something like:
+
+* Prepare **host process resources**:
+
+  * PyMain_InitResourcesConfig()
+  * initialize host process resources
+
+* Populate configs: ``PyMain_InitConfig()`` or something like that.
+
+  * ``PyConfig_Update()``:
+    * (some) env vars → config
+    * **host process resources** → config
+  * Initialize ``PyInterpreterConfig``.
+  * Initialize ``PyMainConfig``.
+
+    * (remaining) env vars → config
+    * argv → config
+    * 3-6. same as above
+
+* Run code or REPL from ``PyMainConfig``.
+
+Such API is more complicated, it requires adding new public structures
+and new public functions. It makes the Python initialization more
+complicated, rather than this PEP tries to unifiy existing APIs and make
+them simpler (the opposite). Having multiple structures for similar
+purpose can lead to duplicate members, similar issue than duplicated
+members between existing ``PyPreConfig`` and ``PyConfig`` structures.
```

Comment:
I think most of the details here are irrelevant to PEP 741.  I recommend summarizing:

```suggestion
`Eric Snow expressed concerns <https://discuss.python.org/t/pep-741-python-configuration-c-api-second-version/45403/27>`_
that this proposal might reinforce with embedders the idea that
initialization is a single monolithic step.  He argued that initialization
involves 5 distinct phases and even suggested that the API should
reflect this explicitly.  Eric proposed that, at the very least, the
implementation of initialization should reflect the phases, in part
for improved code health.  Overall, his explanation has some
similarities with :pep:`432` and :pep:`587`.

Another of Eric's key points relevant to this PEP was that, ideally,
the config passed to ``Py_InitializeFromConfig()`` should be complete
before that function is called, whereas currently initialization
actually modifies the config.

While Eric wasn't necessarily suggesting an alternative to PEP 741,
any proposal to add a granular initialization API around phases is
effectively the opposite of what this PEP is trying to accomplish.
Such API is more complicated, it requires adding new public structures
and new public functions. It makes the Python initialization more
complicated, rather than this PEP tries to unifiy existing APIs and make
them simpler (the opposite). Having multiple structures for similar
purpose can lead to duplicate members, similar issue than duplicated
members between existing ``PyPreConfig`` and ``PyConfig`` structures.

.. XXX Are the concerns still valid about the config currently not
   being complete before ``Py_InitializeFromConfig()`` is called?
```